### ISSUE TITLE
"SupportedDisplayModes" returns no results (Windows DirectX)

### ISSUE THREAD

**Issue by MichaelDePiazzi** [mono/MonoGame#4220]
One of my testers recently tried my game on an older laptop they had. When they tried to set the video resolution though, there were none to select from. Looking into it, `GraphicsAdapter.DefaultAdapter.SupportedDisplayModes` was not returning anything at all.

After some investigation, I found the issue - the laptop only had BGR video modes (specifically `B5G6R5_UNorm`, `B8G8R8A8_UNorm`, and `B8G8R8A8_UNorm_SRgb`). But MonoGame is currently only looking for `R8G8B8A8_UNorm` modes.

Does anyone see any issue with enumerating through all the different `SharpDX.DXGI.Format` values and translating them to the relevant XNA `SurfaceFormat.Color`? (e.g. `Format.B5G6R5_UNorm` -> `SurfaceFormat.Bgr565`, `Format.B8G8R8A8_UNorm` -> `SurfaceFormat.Bgra32`, etc.)

I did notice that WindowGL is currently excluding everything but 32-bit colour modes. I'm not sure if there's a specific reason for this, but I thought it might be relevant and worth mentioning.

PS. I'm more than happy to submit a PR for this. But I just wanted to confirm how this should be resolved first.

**Comment by tomspilman** [comment]
Sure... that sounds reasonable as long as those BGR modes work correctly for the display mode.

**Comment by MichaelDePiazzi** [comment]
Hmmm... well the game works fine (apart from not being able to select a different resolution), so this means that at least the 32-bit `B8G8R8A8_UNorm` format works. I'm not sure about the 16-bit `B5G6R5_UNorm` format though... I'd have to do some tests.

At a minimum, I guess I could just search for `B8G8R8A8_UNorm` modes (and translate it to `SurfaceFormat.Bgra32`) in addition to the current `R8G8B8A8_UNorm` (`SurfaceFormat.Color`) modes. This would at least resolve the issue on this particular laptop.

**Comment by MichaelDePiazzi** [comment]
Well, I just did a quick test run of my game using a `B5G6R5_UNorm` mode and it seems to work correctly.

Testing XNA on my PC, it gives me back `Color` and `Bgr565` modes. Whereas SharpDX gives me modes for several different formats (`R16G16B16A16_Float`, `R10G10B10A2_UNorm`, `R8G8B8A8_UNorm`, `R8G8B8A8_UNorm_SRgb`, `B5G6R5_UNorm`, `B8G8R8A8_UNorm`, `R10G10B10_Xr_Bias_A2_UNorm`, `B8G8R8A8_UNorm_SRgb`). So it seems like XNA is only looking for specific formats.

I also noticed that while my PC has both `R8G8B8A8_UNorm` and `B8G8R8A8_UNorm` modes, XNA seems to only return 1 set of modes for `Color`. So maybe XNA excludes BGR modes if RGB modes are present? I'll try to get my tester to run an XNA test on this laptop and see what modes it gives him.

**Comment by tomspilman** [comment]
>  So it seems like XNA is only looking for specific formats.

Probably always skips floating point formats.  Also probably removes any formats that are essentially duplicates... like BGR modes when RGB is present too.

We just have to test XNA and see if we can guess as to what it is doing internally.

**Comment by MichaelDePiazzi** [comment]
> We just have to test XNA and see if we can guess as to what it is doing internally.

No worries. I've just sent my tester an XNA test app. I'll let you know what we find.

**Comment by MichaelDePiazzi** [comment]
And the results are in - On the laptop with only BGR modes, XNA returns only `Color` and `Bgr565` formats. So it's just treating the "B8G8R8A8" formats as `Color`.

So I'm thinking of just keeping it simple and changing it to also look for `B8G8R8A8_UNorm` and `B5G6R5_UNorm` formats (which will be treated as `Color` and `Bgr565` XNA formats respectively).

This will at least solve the issue with this laptop, and get the results more in line with what XNA is returning. I have a feeling that it won't be necessary to consider any other formats, but I'll write the code in a way that makes it simple to add additional formats later anyway.

What do you think @tomspilman?

**Comment by tomspilman** [comment]
> changing it to also look for  B8G8R8A8_UNorm  and  B5G6R5_UNorm  formats (which will be treated as  Color  and  Bgr565  XNA formats respectively).

Makes sense.  Of course if there is an existing `Color` or `Bgr565` the BGR versions would be ignored to avoid having duplicate entries.

**Comment by MichaelDePiazzi** [comment]
> Of course if there is an existing Color or Bgr565 the BGR versions would be ignored to avoid having duplicate entries.

Yes, that's correct.

I'll try to get a PR for this submitted in the next day or two.

**Comment by mrhelmut** [comment]
Interesting finding! We recently received a crash report related to ```SupportedDisplayModes``` being empty. After further investigation, we noticed that it happens only with some exotic laptop monitors like 1536x864 LCD panels. Even though ```SupportedDisplayModes``` couldn't be enumerated, the game was still working fine.
Since I got that report from an anonymous dump, I haven't been able to trace back the cause. Kudos for that.
